src/test/java/com/wikia/webdriver/common/testnglisteners/InvokeMethodAdapter.java

```diff
@@ -37,25 +40,72 @@ public void afterInvocation(IInvokedMethod method, ITestResult result) {
   }
 
   @Override
-  public void beforeInvocation(IInvokedMethod method, ITestResult testResult) {
-    if (method.isTestMethod() && isTestExcludedFromEnv(method.getTestMethod()
-                                                           .getConstructorOrMethod()
-                                                           .getMethod())) {
-      throw new SkipException("Test can't be run on " + Configuration.getEnv() + " environment");
+  public void beforeInvocation(IInvokedMethod invokedMethod, ITestResult testResult) {
+    Method method = invokedMethod.getTestMethod().getConstructorOrMethod().getMethod();
+    if (invokedMethod.isTestMethod()) {
+      if (isTestExcludedFromEnv(method)) {
+        throw new SkipException("Test can't be run on " + Configuration.getEnv() + " environment");
+      } else if (isTestExcludedFromLang(method)) {
+        throw new SkipException(
+            "Test can't be run with " + Configuration.getWikiLanguage() + " language");
+      }
     }
   }
 
   /**
    * Returns true if test is excluded from running on current test environment
    */
   private boolean isTestExcludedFromEnv(Method method) {
+    String envFromConfig = Configuration.getEnv();
     if (method.isAnnotationPresent(DontRun.class)) {
       String[] excludedEnvs = method.getAnnotation(DontRun.class).env();
 
-      for (String excludedEnv : excludedEnvs) {
-        if (Configuration.getEnv().contains(excludedEnv)) {
-          return true;
-        }
+      if (Arrays.stream(excludedEnvs).anyMatch(e->e.equals(envFromConfig))) {
+        return true;
+      }
+    }
+    if (method.isAnnotationPresent(RunOnly.class)) {
+      String[] includedEnvs = method.getAnnotation(RunOnly.class).env();
+
+      if (hasOnlyDefaultValue(includedEnvs)) {
+        return false;
+      } else if (Arrays.stream(includedEnvs).anyMatch(e->e.equals(envFromConfig))) {
+        return false;
+      } else {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean hasOnlyDefaultValue(String[] includedValues) {
+    return includedValues.length == 1 && includedValues[0].isEmpty();
+  }
+
+  /**
+   * Returns true if test is excluded from running with current language
+   */
+  private boolean isTestExcludedFromLang(Method method) {
+    String langFromConfig = System.getProperty("language") != null ? System.getProperty("language")
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Define a constant instead of duplicating this literal "language" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar-s4/coding_rules#rule_key=squid%3AS1192)